Upgrading to bootstrap5 #4602

salmoneatenbybear opened the pull request:
### Fixes Issue #4602 

### Describe the changes you have made in this PR:

- Migrated the project from Bootstrap 4 to Bootstrap 5.
- Updated Classes and Attributes: Replaced deprecated Bootstrap 4 classes (e.g., .jumbotron, .media, .ml-auto, .float-right) with their Bootstrap 5 equivalents or updated utility classes.
- Data Attribute Renaming: Updated data attributes such as data-toggle and data-target to the new Bootstrap 5 format.
- Bootstrap CDN Links: Replaced any Bootstrap v4 CDN links with the latest Bootstrap v5 versions.
- Utility Class Adjustments: Improved layout and spacing with Bootstrap 5’s updated utility classes.
- Fixed all major layout bugs visible. 

( PLEASE FIND ATTACHED VIDEO OF THE WEBSITE AFTER MIGRATION ) 

This is my second iteration of the PR, addressing all the issues that surfaced during the Bootstrap 4 to 5 migration. I’m grateful for the guidance so far and eager to hear your thoughts on these fixes.

Thank you for the continued support and feedback,
Aditya Singh

cc: @tachyons @tanmoysrt

https://github.com/user-attachments/assets/15d2f354-a486-444d-8ef9-8049bf4f36a6



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Updated social media links in the footer to direct users to official CircuitVerse pages on Facebook, Twitter, LinkedIn, YouTube, and GitHub.

- **UI Enhancements**
  - Transitioned various components to Bootstrap 5, including updates to button classes and modal attributes for improved styling and functionality.
  - Enhanced form layouts with consistent margin classes for better visual alignment.
  - Improved layout and alignment of social icons in the footer.
  - Introduced a new `.zoom-slider` class for better organization of zoom controls.

- **Bug Fixes**
  - Improved responsiveness and styling of alerts and modals throughout the application.

- **Documentation**
  - Updated comments and structure in layout files for better clarity and organization.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tachyons commented:
@salmoneatenbybear It seems that you accidentally added changes to git sub module on this PR and it causing CI to fail. Also could you check the issues raised by the coder rabbit ? Great job so far

salmoneatenbybear commented:
@tachyons absolutely will look into it, thankyou

salmoneatenbybear commented:
@tachyons I have reset the submodule changes, and updated the PR.

salmoneatenbybear commented:
@tachyons I have looked into the changes suggested by coder rabbit and made those changes aswell, please review if you get time.

salmoneatenbybear commented:
@tachyons I have tried to go to an earlier hash and hard reset from there in order to reinstate the sub modules, i haven't changed much in there just updated the QuickButtonMobile.vue and QuickButton.vue so that the ui remains stable across all devices, I am not sure why the CI keeps failing.

tachyons commented:
<img width="1168" alt="Screenshot 2024-10-20 at 6 47 04 PM" src="https://github.com/user-attachments/assets/b2719e33-a039-4c24-88e5-b5ce1682dd4b">
CI is failing because of this 


Try https://how.wtf/reverting-changes-to-git-submodules.html

salmoneatenbybear commented:
> <img alt="Screenshot 2024-10-20 at 6 47 04 PM" width="1168" src="https://private-user-images.githubusercontent.com/3112976/378190816-b2719e33-a039-4c24-88e5-b5ce1682dd4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk0NDE1OTUsIm5iZiI6MTcyOTQ0MTI5NSwicGF0aCI6Ii8zMTEyOTc2LzM3ODE5MDgxNi1iMjcxOWUzMy1hMDM5LTRjMjQtODhlNS1iNWNlMTY4MmRkNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjBUMTYyMTM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjQ0YWVhN2EzYzJiODMzNTFlM2I2M2U2YjJlZDY1OTFiZjQ2YmVlMTcxMjNiZTljNGI3YTZjMmYzMjI1OGVkZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.nX_aiAiI7LoNDoIPueFgLbdOcEepwnsMJwtb4XHg8kE"> CI is failing because of this
> Try https://how.wtf/reverting-changes-to-git-submodules.html

Have tried to resolve by reverting the changes

tachyons commented on db/schema.rb:
```diff
@@ -92,7 +92,7 @@
   create_table "assignments", force: :cascade do |t|
     t.string "name"
-    t.datetime "deadline", null: false
+    t.datetime "deadline", precision: nil, null: false
```
Database changes are unrelated, please remove this.

salmoneatenbybear replied:
alright

tachyons replied:
@salmoneatenbybear This isn't address yet

tachyons commented on apidoc/v1/index.html:
```diff
@@ -215,173 +215,173 @@
         </div>
       <ul id="toc" class="toc-list-h1">
           <li>
-            <a href="#overview" class="toc-h1 toc-link" data-title="Overview">Overview</a>
```
Do not change this file. This a autogenerated file from docs/ dir

salmoneatenbybear replied:
alright doing this now, done for db/schema.rb

tachyons commented:
@salmoneatenbybear specs are failing for groups, could you check that as well ? Like adding and removing members to the group ?

salmoneatenbybear commented:
> @salmoneatenbybear specs are failing for groups, could you check that as well ? Like adding and removing members to the group ?

Sure absolutely, will do that asap

salmoneatenbybear commented:
@tachyons so i figured out why one of the CI checks was failing ( rest which were failing before due to the submodule changes are resolved ), its failing because : 

- The "delete" button (groups-assignment-delete-button) is missing from the assignment views.
- There are parameter issues like an empty or missing group param, causing the tests to fail.

i havent made any changes related to this in this pr so i am not sure why this is occuring, i can try to fix this but for that i will need to add the missing "delete" button in the `show.html.erb` file under the mentor access section and link it to the destroy action.

tachyons commented:
It could be related to bootstrap JavaScript. Try if the feature works on
your branch like in the master branch

On Mon, 21 Oct, 2024, 3:35 pm Aditya Singh, ***@***.***>
wrote:

> @tachyons <https://github.com/tachyons> so i figured out why one of the
> CI checks was failing ( rest which were failing before due to the submodule
> changes are resolved ), its failing because :
>
>    - The "delete" button (groups-assignment-delete-button) is missing
>    from the assignment views.
>    - There are parameter issues like an empty or missing group param,
>    causing the tests to fail.
>
> i havent made any changes related to this in this pr so i am not sure why
> this is occuring, i can try to fix this but for that i will need to add the
> missing "delete" button in the show.html.erb file under the mentor access
> section and link it to the destroy action.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CircuitVerse/CircuitVerse/pull/5117#issuecomment-2426218350>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXYAEB7YXKEHTR2J2UTPUTZ4TGYJAVCNFSM6AAAAABQIM4CJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMRWGIYTQMZVGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

salmoneatenbybear commented:
> It could be related to bootstrap JavaScript. Try if the feature works on
> your branch like in the master branch
> 
> On Mon, 21 Oct, 2024, 3:35 pm Aditya Singh, ***@***.***>
> wrote:
> 
> > @tachyons <https://github.com/tachyons> so i figured out why one of the
> > CI checks was failing ( rest which were failing before due to the submodule
> > changes are resolved ), its failing because :
> >
> >    - The "delete" button (groups-assignment-delete-button) is missing
> >    from the assignment views.
> >    - There are parameter issues like an empty or missing group param,
> >    causing the tests to fail.
> >
> > i havent made any changes related to this in this pr so i am not sure why
> > this is occuring, i can try to fix this but for that i will need to add the
> > missing "delete" button in the show.html.erb file under the mentor access
> > section and link it to the destroy action.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/CircuitVerse/CircuitVerse/pull/5117#issuecomment-2426218350>,
> > or unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AAXYAEB7YXKEHTR2J2UTPUTZ4TGYJAVCNFSM6AAAAABQIM4CJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMRWGIYTQMZVGA>
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
> 

Alright I will try it and keep you updated.

salmoneatenbybear commented:
@tachyons CI is failing due to Capybara not finding elements on the page and missing parameters, trying to fix them...

salmoneatenbybear commented:
@tachyons the CI was failing beacause capybara is trying to find the bootstrap 4 id which has been changed when migrating to bootstrap 5, so the tests need to be changed with the appropriate bootstrap 5 id which i replaced the bootstrap 4 ids with, should i make the changed to the tests ?

tachyons commented:
Yes please

salmoneatenbybear commented:
@tachyons pls review if you find time. Thankyou!

tachyons commented on apidoc/v1/index.html:
```diff
@@ -3689,4 +3689,4 @@ <h3 id='possible-exceptions-3'>Possible exceptions</h3>
     </div>
   </body>
-</html>
+</html>
```
@salmoneatenbybear This file is still in the diff. 

You can 

`git checkout master apidoc/v1/index.html` to reset

tachyons commented on app/assets/stylesheets/components.scss:
```diff
@@ -285,7 +285,23 @@
+  align-items: center; // Vertically align both the text and button
+}
+
+.btn-close {
+  margin-right: 10px; 
+  background: none;
+  border: none;
+  color: black;
+  font-size: 1.5rem;
+  opacity: 1;
+  font-weight: bold;
+  padding: 0.5rem;
+  
```
@salmoneatenbybear Why this wasn't addressed  ?

salmoneatenbybear replied:
Apologies for the oversight, I missed this suggestion earlier. I have now removed the custom .btn-close styling, The additional &times; symbol that was causing the duplicate "X" has also been removed from the code.

The custom styling is now replaced with Bootstrap 5's default styling 

@tachyons

tachyons commented on app/assets/stylesheets/groups.scss:
```diff
@@ -19,9 +19,23 @@
 
 .groups-primary-button {
-  padding: 12px 35px;
+  font-size: 14px;  // Make the text smaller for a compact look
+  padding: 10px 25px;  // Adjust padding to make the button smaller
+  border-radius: 5px;  // Rounded corners
+  float: right;  // Align the button to the right
+  margin-left: 10px;  // Add space between the button and text
+  transition: background-color 0.3s ease;  // Smooth transition on hover
```
@salmoneatenbybear Please don't close a thread without addressing it or by leaving a comment about why the approach is not considered

salmoneatenbybear replied:
addressed, removed float : right , and made sure the UI is not broken.

salmoneatenbybear replied:
@tachyons apolgies for closing the thread without addressing

tachyons commented on app/assets/stylesheets/groups.scss:
```diff
@@ -19,9 +19,23 @@
+  transition: background-color 0.3s ease;  // Smooth transition on hover
 }
 
+.groups-section {
+  display: flex;
+  justify-content: space-between;  // Ensures the button is on the right side
+  align-items: center;
+  margin-bottom: 15px;  // Add space between sections
+}
```
@salmoneatenbybear Please don't close a thread without addressing it or by leaving a comment about why the approach is not considered

tachyons commented on app/assets/stylesheets/users.scss:
```diff
@@ -109,6 +109,12 @@
+  margin-bottom: 20px; /* Add some spacing between the icons and the text below */
+}
+
+
```
Remove extra spaces

salmoneatenbybear replied:
removed

tachyons commented on app/component/social_links_component.rb:
```diff
@@ -3,11 +3,11 @@
-      { name: "linkedin", url: "/linkedin", logo: "logos/linkedin-logo.png" },
-      { name: "youtube", url: "/youtube", logo: "logos/youtube-logo.png" },
-      { name: "github", url: "/github", logo: "logos/github-logo-circle.png" }
+      { name: "facebook", url: "https://www.facebook.com/CircuitVerse", logo: "logos/facebook-logo.png" },
+      { name: "twitter", url: "https://twitter.com/CircuitVerse", logo: "logos/twitter-x.png" },
+      { name: "linkedin", url: "https://www.linkedin.com/company/circuitverse", logo: "logos/linkedin-logo.png" },
+      { name: "youtube", url: "https://www.youtube.com/c/CircuitVerse", logo: "logos/youtube-logo.png" },
+      { name: "github", url: "https://github.com/CircuitVerse", logo: "logos/github-logo-circle.png" }
```
Not needed for now as we have bultin routing for third party links

tachyons commented on package.json.save:
```diff
@@ -0,0 +1,81 @@
+{
```
@salmoneatenbybear Remove this file

salmoneatenbybear replied:
removed

tachyons commented on package.json.save:
```diff
@@ -0,0 +1,81 @@
+    "jquery-ujs": "^1.2.3",
+    "jsdoc-export-default-interop": "^0.3.1",
+    "opencollective-postinstall": "^2.0.3",
+<<<<<<< HEAD
+    "popper.js": "^1.16.1",
+    "sass": "^1.80.3",
+=======
+    "@popperjs/core": "^2.11.8",
+    "sass": "^1.79.3",
+>>>>>>> ed0196d8 (refactor: upgrade Bootstrap from v4 to v5)
```
@salmoneatenbybear check this

salmoneatenbybear replied:
addressed

salmoneatenbybear commented:
I have reset the apidoc file,

removed the whitespace

and also removed the package.json.save from the changes.

if anything else needs changes, updates/modifications, etc; pls do let me know, would be happy to make them.

Thankyou for your guidance @tachyons

tachyons commented:
<img width="938" alt="Screenshot 2024-10-25 at 10 31 55 PM" src="https://github.com/user-attachments/assets/65fa5621-1645-4cd6-a4a2-871869d70714">

Could you address the open issues raised by the @coderabbit ?

salmoneatenbybear commented:
> <img width="938" alt="Screenshot 2024-10-25 at 10 31 55 PM" src="https://github.com/user-attachments/assets/65fa5621-1645-4cd6-a4a2-871869d70714">
> 
> Could you address the open issues raised by the @coderabbit ?

Sure absolutely, will do that

salmoneatenbybear commented on simulator/src/css/main.stylesheet.css:
```diff
@@ -1603,19 +1603,25 @@ input:checked + .slider:before {
 }
 
 .zoom-slider {
+    display: flex;
+    align-items: center;
```
I have refactored the slider using Bootstrap's utility classes as much as possible. However, due to the specific UI requirements of the slider, it wasn't fully achievable using only Bootstrap's utility classes without compromising the design and functionality. The retained CSS is the minimal amount necessary to ensure the UI remains consistent and behaves as expected.

salmoneatenbybear commented on app/views/commontator/comments/_delete_comment_confirmation_modal.html.erb:
```diff
@@ -3,12 +3,12 @@
-      <div class="modal-footer" data-dismiss="modal">
+      <div class="modal-footer" data-bs-dismiss="modal">
         <%= link_to "#",
                     method: :put,
```
done

salmoneatenbybear commented on app/views/layouts/_header.html.erb:
```diff
@@ -46,14 +46,14 @@
             <a class="dropdown-item" href="<%= user_groups_path(current_user) %>"><%= t("layout.header.my_groups") %></a>
             <div class="dropdown-divider"></div>
-            <a class="dropdown-item" rel="nofollow" data-method="delete" href="<%= destroy_user_session_path %>"><%= t("sign_out") %></a>
+            <a class="dropdown-item" rel="nofollow" data-bs-method="delete" href="<%= destroy_user_session_path %>"><%= t("sign_out") %></a>
```
done

salmoneatenbybear commented on simulator/src/css/main.stylesheet.css:
```diff
@@ -1603,19 +1603,25 @@ input:checked + .slider:before {
+    font-size: 18px;
+    width: 40px;
+    height: 40px;
+    border-radius: 5px;
```
done as much as possible.

salmoneatenbybear commented on app/views/groups/show.html.erb:
```diff
@@ -29,38 +29,51 @@
-  <% if @group.group_members.member.blank? %>
-      <span class="null-text"><%= t("groups.show.no_members_html") %></span>
+  <% if policy(@group).admin_access? %>
+    <button type="button" class="btn primary-button groups-primary-button" 
```
addressed

salmoneatenbybear commented on app/component/social_links_component.html.erb:
```diff
@@ -1,4 +1,4 @@
-<div class="row footer-social-icons-row">
+<div class="footer-social-icons-container">
   <% social_links.each do |link| %>
     <%= link_to link[:url], class: "bounce-out-on-hover", target: "_blank", data: { tooltip: link[:name] } do %>
```
I refactored the alert auto-hide script by removing the jQuery dependency and implementing a modern JavaScript approach. The new script listens for the DOMContentLoaded event and uses native DOM manipulation methods (querySelector and style.display). The logic remains the same: hiding the alert after 3 seconds, but the updated code avoids the overhead of using jQuery when it's not necessary. @tachyons please check if the code has any bugs.

salmoneatenbybear replied:
Done

salmoneatenbybear commented on app/views/layouts/application.html.erb:
```diff
@@ -27,13 +31,19 @@
           setTimeout(() => {document.querySelector(".alert").style.display == "none";}, 3000);
         }
       })
     </script>
```
I refactored the alert auto-hide script by removing the jQuery dependency and implementing a modern JavaScript approach. The new script listens for the DOMContentLoaded event and uses native DOM manipulation methods (querySelector and style.display). The logic remains the same: hiding the alert after 3 seconds, but the updated code avoids the overhead of using jQuery when it's not necessary.

salmoneatenbybear commented on app/views/layouts/application.html.erb:
```diff
@@ -3,37 +3,50 @@
     </script>
+
+
+    <!-- Load Popper.js (required for Bootstrap 5 dropdowns and tooltips) -->
+    <script src="https://cdn.jsdelivr.net/npm/@popperjs/core@2.11.7/dist/umd/popper.min.js"></script>
+
+    <!-- Load Bootstrap JS -->
+    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.0/dist/js/bootstrap.min.js"></script>
```
I've kept the CDN links for Bootstrap and Popper.js as a fallback for now. Removing them would require importing these libraries directly in application.js, and then configuring the asset pipeline or Webpack to handle the JavaScript compilation. This process involves additional steps to ensure everything is compiled and loaded properly.

salmoneatenbybear commented on app/views/users/shared/_links.html.erb_spec.rb:
```diff
@@ -0,0 +1,129 @@
+    allow(view).to receive(:t).with("groups.use_alternative").and_return("Use alternative")
+    allow(view).to receive(:t).with("users.shared.login_with_sso").and_return("Login with SSO")
+    allow(view).to receive(:t).with("users.shared.resend_email").and_return("Didn't receive confirmation instructions?")
+  end
```
I think i would need a little guidnance from the maintainer for this, untill then I would refrain from doing this in order to not break any existing things.

salmoneatenbybear commented:
@tachyons 

The ruby on rails test CI failure seems to be caused by GitHub's permission restrictions on workflows triggered from forked repositories.

Specifically, the reviewdog tool in the CI pipeline requires write permissions to post review comments on pull requests. However, when a PR comes from a fork (like mine), GitHub limits the GITHUB_TOKEN permissions for security reasons, leading to a 403 Resource not accessible by integration error.

I believe this is why the CI check failed, not due to issues with the code itself. Could you please advise on how we should proceed? I'm happy to make any necessary adjustments.

Thank you for your time and guidance.

salmoneatenbybear commented:
@tachyons I have closed this PR due to the submodule issue which was causing the CI to fail, i will be creating a new PR with the same changes so that the CI passes.

salmoneatenbybear commented:
superseded by #5135
